### PR TITLE
✨ Authentication with LDAP

### PR DESCRIPTION
**Pull request by wizzytod**
Users can sign in/sign up with LDAP credentials using the same login form.

Sign in with LDAP

### CODE REVIEW

**Comment by niwinz** [comment]
at quick code view, looks awesome, nice job

File: backend/deps.edn [hunk 1]
```diff
@@ -53,6 +53,8 @@
   com.draines/postal {:mvn/version "2.0.3"
                       :exclusions [commons-codec/commons-codec]}
 
+  org.clojars.pntblnk/clj-ldap {:mvn/version"0.0.16"}
```
Comment (madmath03) [inline backend/deps.edn]: @wizzytod should we use [puppetlabs/clj-ldap](https://github.com/puppetlabs/clj-ldap) instead ?

I see that [pntblnk/clj-ldap](https://github.com/pauldorman/clj-ldap) has not be updated for 3 years, but Puppet forked it and updates quite recently.
Might be safer for the long term. I let you check and validate if switch is possible.

File: backend/src/uxbox/config.clj [hunk 2]
```diff
@@ -43,8 +43,22 @@
...
+   ;:ldap-bind-dn "cn=admin,dc=ldap,dc=mysupercompany,dc=com"
+   ;:ldap-bind-password "verysecure"
+   ;:ldap-auth-ssl false
+   ;:ldap-auth-starttls true
+   ;:ldap-auth-base-dn "ou=People,dc=ldap,dc=mysupercompany,dc=com"
+   ;:ldap-auth-user-query "(|(uid=$username)(mail=$username))"
+   ;:ldap-auth-username-attribute "uid"
+   ;:ldap-auth-email-attribute "mail"
+   ;:ldap-auth-fullname-attribute "displayname"
+   ;:ldap-auth-avatar-attribute "jpegPhoto"
```
Comment (madmath03) [inline backend/src/uxbox/config.clj]: I think we can safely set default values like query and attributes that would match most cases.

```suggestion
   :debug-humanize-transit true

   ;; LDAP auth disabled by default. Set ldap-auth-host to enable
   ;:ldap-auth-host "ldap.mysupercompany.com"
   ;:ldap-auth-port 389
   :ldap-auth-version "3"
   ;:ldap-bind-dn "cn=admin,dc=ldap,dc=mysupercompany,dc=com"
   ;:ldap-bind-password "verysecure"
   ;:ldap-auth-ssl false
   ;:ldap-auth-starttls false
   ;:ldap-auth-base-dn "ou=People,dc=ldap,dc=mysupercompany,dc=com"
   :ldap-auth-user-query "(|(uid=$username)(mail=$username))"
   :ldap-auth-username-attribute "uid"
   :ldap-auth-email-attribute "mail"
   :ldap-auth-fullname-attribute "displayName"
   :ldap-auth-avatar-attribute "jpegPhoto"})
```

File: backend/src/uxbox/config.clj [hunk 3]
```diff
@@ -78,6 +92,21 @@
 (s/def ::google-client-id ::us/string)
 (s/def ::google-client-secret ::us/string)
 
+(s/def ::ldap-auth-host ::us/string)
+(s/def ::ldap-auth-port ::us/integer)
+(s/def ::ldap-auth-version ::us/string)
```
Comment (madmath03) [inline backend/src/uxbox/config.clj]: I see that `clj-ldap` does not support configuration of LDAP version, so we can remove `ldap-auth-version`.
```suggestion
```

File: frontend/resources/locales.json [hunk 4]
```diff
@@ -107,6 +107,15 @@
       "ru" : "Вход"
     }
   },
+  "auth.login-with-ldap-submit-label" : {
+    "used-in" : [ "src/uxbox/main/ui/auth/login.cljs:108" ],
+    "translations" : {
+      "en" : "Sign in with LDAP",
+      "fr" : "",
+      "es" : "",
```
Comment (madmath03) [inline frontend/resources/locales.json]:
```suggestion
      "fr" : "Se connecter via LDAP",
      "es" : "Entrar con LDAP",
```

**Review by madmath03 — CHANGES_REQUESTED** [review]
A few questions for you @wizzytod , but overall great work.

I'm gonna send an update of the management guide documentation to complete this.

File: backend/deps.edn (hunk 1 above)
Comment (wizzytod) [reply backend/deps.edn]: I agree, adopted code to slightly different API. thanks!

**Comment by niwinz** [comment]
Is this ready to merge?

**Comment by niwinz** [comment]
Additionally, would be awesome to have a concrete (and if possible easy/simple) instructions to test this. Maybe a docker container on the devenv with ldap provider with default data or something similar.

**Comment by madmath03** [comment]
> Is this ready to merge?

Unless you see anything wrong during review, then yes it's ready to be merged.

**Comment by madmath03** [comment]
> Additionally, would be awesome to have a concrete (and if possible easy/simple) instructions to test this. Maybe a docker container on the devenv with ldap provider with default data or something similar.

Would be nice indeed, but might take some time to get a local OpenLDAP working (it can easily get tricky to configure).
We will have a look to what we can do for that.

**Comment by niwinz** [comment]
Not worry, this is not blocking for merge this, but in order to be able to maintain this without accidental breaking, we need a way check that integration works. 

Honestly, I know nothing about LDAP, so right now i can only review the code but can't test it. If in a future release we change something in the auth subsystem, without a test environment for ldap we can leave accidentally the LDAP broken until some user that uses it, detect the fail...

**Comment by niwinz** [comment]
Merged manually

**Pull request closed without merging**